src/sources/kafka.rs

```diff
@@ -370,105 +411,559 @@ impl SourceConfig for KafkaSourceConfig {
 
 async fn kafka_source(
     config: KafkaSourceConfig,
-    consumer: StreamConsumer<CustomContext>,
+    consumer: StreamConsumer<KafkaSourceContext>,
     decoder: Decoder,
-    mut shutdown: ShutdownSignal,
-    mut out: SourceSender,
-    acknowledgements: bool,
+    out: SourceSender,
+    shutdown: ShutdownSignal,
+    eof: bool,
     log_namespace: LogNamespace,
 ) -> Result<(), ()> {
+    let span = info_span!("kafka_source");
     let consumer = Arc::new(consumer);
-    let (finalizer, mut ack_stream) =
-        OrderedFinalizer::<FinalizerEntry>::maybe_new(acknowledgements, Some(shutdown.clone()));
-    let finalizer = finalizer.map(Arc::new);
-    if let Some(finalizer) = &finalizer {
-        consumer
-            .context()
-            .finalizer
-            .set(Arc::clone(finalizer))
-            .expect("Finalizer is only set once");
+
+    consumer
+        .context()
+        .consumer
+        .set(Arc::downgrade(&consumer))
+        .expect("Error setting up consumer context.");
+
+    let (callback_sender, callback_rx) = mpsc::unbounded_channel();
+    consumer
+        .context()
+        .callbacks
+        .set(callback_sender)
+        .expect("Error setting up consumer callback channel.");
+
+    // EOF signal allowing the coordination task to tell the kafka client task when all partitions have reached EOF
+    let (eof_tx, eof_rx) = if eof {
+        let (tx, rx) = oneshot::channel::<()>();
+        (Some(tx), Some(rx))
+    } else {
+        (None, None)
+    };
+
+    let coordination_task = {
+        let span = span.clone();
+        let consumer = Arc::clone(&consumer);
+        let drain_timeout_ms = match config.drain_timeout_ms {
+            Some(d) => Duration::from_millis(d),
+            None => config.session_timeout_ms / 2,
+        };
```

Comment:
Also FWIW We also have `Option::map_or`
```suggestion
        let drain_timeout_ms = config.drain_timeout_ms.map_or(config.session_timeout_ms / 2, Duration::from_millis);
```